tools/oss-fuzz.sh

```diff
@@ -56,6 +64,17 @@ fi
 
 ninja -v -C "$build" fuzzers
 
+# Compressed BCD files are kept in test/test-bcd so let's unpack them
+# and populate the directory the script uses to generate seed corpora
+rm -rf test/fuzz/fuzz-bcd/
+mkdir -p test/fuzz/fuzz-bcd/
+
+# shellcheck disable=SC2016
+# Single quotes are used here intentionally to let sh expand $
+# when it's run by xargs
+find "test/test-bcd" -type f -name '*.zst' -print0 |
+xargs -0 -I%%FILE%% sh -x -c 'unzstd %%FILE%% -o "test/fuzz/fuzz-bcd/$(basename %%FILE%% .zst)"'
```

Comment:
Something like this is much nicer:
```shell
bcd=$(mktemp -d)
for i in test/test-bcd/*.zst; do
     unzstd "$i" -o "$bcd/$(basename "${i%.zst}")";
done
zip -jqr "$OUT/fuzz-bcd_seed_corpus.zip" "$bcd"
rm -rf "$bcd"
```

Reply:
I don't think `for` can handle directories that public OSS-Fuzz corpora bring with them so I don't think `find` should be replaced with it. Other than that, I agree the script should put those files in a temporary directory.

Ideally, the bcd parser should have been integrated properly into the test suite though to avoid kludges like that. In its current form it isn't compatible with the regression test suite.

Reply:
> I don't think `for` can handle directories that public OSS-Fuzz corpora bring with them so I don't think `find` should be replaced with it.

What exactly makes you think this for is unsafe?

Reply:
> Ideally, the bcd parser should have been integrated properly into the test suite though to avoid kludges like that. In its current form it isn't compatible with the regression test suite.

What exactly makes it incompatible?

Reply:
> What exactly makes you think this for is unsafe?

I'm not saying it's unsafe. I'm saying that if directories are added to `test/test-bcd` that for loop would just skip them because "test/test-bcd/*.zst" can't expand, for example, "test/test-bcd/regressions/crash-...."

> What exactly makes it incompatible?

The regression testsuite expects files to be kept in `test/fuzz/fuzz-*`

Reply:
> I'm not saying it's unsafe. I'm saying that if directories are added to test/test-bcd that for loop would just skip them because "test/test-bcd/*.zst" can't expand, for example, "test/test-bcd/regressions/crash-...."

bash has `shopt -s globstar` for that.

Reply:
I think the problem is that most `for` bash loops in the systemd codebase (and I assume everywhere else) don't expect it to be enabled by default and I certainly wouldn't want to start rewriting those scripts to just replace `find` with `for`.

Reply:
You don't have to change anything. no shell file has `**` in it. Even if they had, you don't need to change unrelated scripts.

Reply:
I'm still not sure why
```
shopt -s globstar
for i in test/test-bcd/**.zst; do
     unzstd "$i" -o "$bcd/$(basename "${i%.zst}")";
done
```
is better than
```
find "test/test-bcd" -type f -name '*.zst' -print0 |
xargs -0 -I%%FILE%% sh -x -c 'unzstd %%FILE%% -o "'"$bcd"'/$(basename %%FILE%% .zst)"'
```
though :-) Since it's maintained by @mrc0mmand mostly @mrc0mmand I wonder what you'd prefer here?

Reply:
It's more readable (less quote-nesting) and doesn't involve xargs. Imho, shell (bash) features should always be preferred to some crazy cmd pipeline.

Btw, the code as-is doesn't even work if subdirs were involved. Or rather it flattens the subdir structure, making the whole thing moot to begin with (and risk overriding files, which could hide real issues). I don't think subdirs are likely to be ever needed (for bcd tests).

Reply:
> the code as-is doesn't even work if subdirs were involved

It doesn't work in general. But files kept in those "regression" directories are named according to their hash sums and collisions there are negligible.

Anyway, since it's mostly my use case I can keep it locally so I'll replace find with for loop without `shopt` though because `test/test-bcd` is as flat as it can get.

Reply:
FWIW @medhefgo if you're interested I can add your email address to https://github.com/google/oss-fuzz/blob/master/projects/systemd/project.yaml. It should be a gmail address though due to https://google.github.io/oss-fuzz/faq/#why-do-you-require-a-google-account-for-authentication

Reply:
find was replaced with for

Reply:
> I'm still not sure why
> 
> ```
> shopt -s globstar
> for i in test/test-bcd/**.zst; do
>      unzstd "$i" -o "$bcd/$(basename "${i%.zst}")";
> done
> ```
> 
> is better than
> 
> ```
> find "test/test-bcd" -type f -name '*.zst' -print0 |
> xargs -0 -I%%FILE%% sh -x -c 'unzstd %%FILE%% -o "'"$bcd"'/$(basename %%FILE%% .zst)"'
> ```
> 
> though :-) Since it's maintained by @mrc0mmand mostly @mrc0mmand I wonder what you'd prefer here?

I can throw in a third variant to make things more complicated ;)

```bash
while read -r file; do
    unzstd "$file" -o "$bcd/$(basename "${file%.zst}")";
done < <(find "test/test-bcd" -type f -name '*.zst') 
```

This seems to be somewhere in the middle of both proposed variants. So far I kept avoiding `shopt`s, since they are, let's say, not that common, so it might make the shell code less understandable (at least on the first glance), but in the end I don't really care which variant eventually "wins".